File: boj_6186.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    public static void main(String args[]) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int[] dr = {-1, 0, 1, 0};
        int[] dc = {0, -1, 0, 1};
        int r = Integer.parseInt(st.nextToken());
        int c = Integer.parseInt(st.nextToken());
        int[][] map = new int[r][c];
        for (int i = 0; i < r; i++) {
            String str = br.readLine();
            for (int j = 0; j < c; j++) {
                if (str.charAt(j) == '#') map[i][j] = 1;
            }
        }
        int cnt = 0;
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                if (map[i][j] == 1) {
                    cnt++;
                    map[i][j] = 0;
                    for (int k = 0; k < 4; k++) {
                        int nr = i + dr[k];
                        int nc = j + dc[k];
                        if (nr < 0 || nr > r - 1 || nc < 0 || nc > c - 1) continue;
                        map[nr][nc] = 0;
                    }
                }
            }
        }
        System.out.println(cnt);
    }
}

File: boj_6186.cpp
#include <string.h>

#include <iostream>

using namespace std;

int main() {
    int dr[4] = {-1, 0, 1, 0};
    int dc[4] = {0, -1, 0, 1};
    int r, c;
    string str;
    cin >> r >> c;
    int map[r][c] = {0};
    for (int i = 0; i < r; i++) {
        cin >> str;
        for (int j = 0; j < c; j++) {
            if (str[j] == '#') map[i][j] = 1;
        }
    }
    int cnt = 0;
    for (int i = 0; i < r; i++) {
        for (int j = 0; j < c; j++) {
            if (map[i][j] == 1) {
                cnt++;
                map[i][j] = 0;
                for (int k = 0; k < 4; k++) {
                    int nr = i + dr[k];
                    int nc = j + dc[k];
                    if (nr < 0 || nr > r - 1 || nc < 0 || nc > c - 1) continue;
                    map[nr][nc] = 0;
                }
            }
        }
    }
    cout << cnt << endl;
    return 0;
}
